shell/platform/android/test/io/flutter/plugin/platform/PlatformPluginTest.java

```diff
@@ -32,4 +46,45 @@ public void itIgnoresNewHapticEventsOnOldAndroidPlatforms() {
     // SELECTION_CLICK haptic response is only available on "LOLLIPOP" (21) and later.
     platformPlugin.vibrateHapticFeedback(PlatformChannel.HapticFeedbackType.SELECTION_CLICK);
   }
+
+  @Test
+  public void platformPlugin_hasStringsMessage() {
+    MethodChannel rawChannel = mock(MethodChannel.class);
+    FlutterJNI mockFlutterJNI = mock(FlutterJNI.class);
+    DartExecutor dartExecutor = new DartExecutor(mockFlutterJNI, mock(AssetManager.class));
+    PlatformChannel fakePlatformChannel = new PlatformChannel(dartExecutor);
+    PlatformChannel.PlatformMessageHandler mockMessageHandler = mock(PlatformChannel.PlatformMessageHandler.class);
+    fakePlatformChannel.setPlatformMessageHandler(mockMessageHandler);
+    Boolean returnValue = true;
+    when(mockMessageHandler.clipboardHasStrings()).thenReturn(returnValue);
+    MethodCall methodCall = new MethodCall("Clipboard.hasStrings", null);
+    MethodChannel.Result mockResult = mock(MethodChannel.Result.class);
+    fakePlatformChannel.parsingMethodCallHandler.onMethodCall(methodCall, mockResult);
+
+    JSONObject expected = new JSONObject();
+    try {
+      expected.put("value", returnValue);
+    } catch (JSONException e) {
+    }
+    verify(mockResult).success(Matchers.refEq(expected));
+  }
+
+  @Test
+  public void platformPlugin_hasStrings() {
+    View fakeDecorView = mock(View.class);
+    Window fakeWindow = mock(Window.class);
+    when(fakeWindow.getDecorView()).thenReturn(fakeDecorView);
+    Activity fakeActivity = mock(Activity.class);
+    when(fakeActivity.getWindow()).thenReturn(fakeWindow);
+    PlatformChannel fakePlatformChannel = mock(PlatformChannel.class);
+    PlatformPlugin platformPlugin = new PlatformPlugin(fakeActivity, fakePlatformChannel);
+
+    ClipboardManager clipboardManager =
+        RuntimeEnvironment.application.getSystemService(ClipboardManager.class);
+    clipboardManager.setText("iamastring");
+    assertTrue(platformPlugin.mPlatformMessageHandler.clipboardHasStrings());
```

Comment:
@gaaclarke Thanks for all the help with the tests.  The first test above works, but this one is failing because the engine is unable to use the clipboard internally.

I thought that my use of ShadowClipboardManager would solve this problem, but maybe I'm doing this wrong?  I got this from [InputConnectionAdaptorTest](https://github.com/flutter/engine/blob/master/shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java).

The engine is unable to get a ClipboardManager here (it's null):

https://github.com/flutter/engine/blob/f1855fb5a4b8f7219af1ed7597e1ea2a3f0c90d2/shell/platform/android/io/flutter/plugin/platform/PlatformPlugin.java#L272-L274

Reply:
That's weird, you can try moving the ClipboardManager to the top of the test?  I'm not sure if there is a temporal problem here.  I don't see a difference between what you are doing and the other test.  Worse case scenario you can make your own interface on top of the ClipboardManager that the PlatformPlugin uses.  It would be nice if we could get this working though.  Sorry, gotta run to a meeting.

Reply:
Thanks, I think I'm on to something with that.  I'll post back later with results.

Reply:
I believe it's working, thank you!  Let me know if it looks alright. I did have to make one protected method public at some point.